FCMExternalProberTarget/app/src/androidTest/java/com/google/firebase/messaging/testing/fcmexternalprobertarget/FCMReceiveServiceInstrumentedTest.java

```diff
@@ -0,0 +1,89 @@
+/*
+ * Copyright 2020 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.firebase.messaging.testing.fcmexternalprobertarget;
+
+import android.content.Context;
+
+import androidx.test.platform.app.InstrumentationRegistry;
+
+import com.google.firebase.messaging.RemoteMessage;
+
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+
+import static org.junit.Assert.*;
+
+import java.io.File;
+import java.io.IOException;
+import java.time.Clock;
+import java.time.Instant;
+import java.time.ZoneId;
+import java.util.Collections;
+import java.util.Map;
+import java.util.Scanner;
+
+public class FCMReceiveServiceInstrumentedTest {
```

Comment:
I had to look up why these tests appeared duplicated.

"Instrumented unit tests are tests that run on physical devices and emulators, and they can take advantage of the Android framework APIs and supporting APIs, such as AndroidX Test. Instrumented tests provide more fidelity than local unit tests, but they run much more slowly. Therefore, we recommend using instrumented unit tests only in cases where you must test against the behavior of a real device."

legit question, does this test against the behavior of a real device?

Reply:
In the case of the FCMReceiveService test, the external storage of the emulated device is used for storing the log file. In the case of the LogReceiver test, an actual textView is written to for logging operations. I'm not certain if the latter uses the device effectively enough to warrant a test with an emulator. My original rationale for writing such a test was to utilize the actual implementations of Android libraries, which may expose issues that the mocks might not.